lib/escalate.rb

```diff
@@ -1,8 +1,78 @@
 # frozen_string_literal: true
 
+require "active_support"
+require "active_support/core_ext"
+require "contextual_logger"
+
 require_relative "escalate/version"
+require_relative "escalate/mixin"
 
 module Escalate
   class Error < StandardError; end
-  # Your code goes here...
+
+  class << self
+    # Logs and escalated an exception
+    #
+    # @param [Exception] exception
+    #   The exception that was raised and needs to be escalated
+    #
+    # @param [String] message
+    #   An additional message about what was happening at the time of the exception
+    #
+    # @param [Logger] logger
+    #   The logger object to use when logging the exception
+    #
+    # @param [Hash] context
+    #   Any additional context to be tied to the escalation
+    def escalate(exception, message, logger, **context)
+      error_message = <<~EOS
+        [Escalate] #{message} (#{context.inspect})
+          #{exception.class.name}: #{exception.message}
+          #{exception.backtrace.join("\n")}
+      EOS
+
+      if logger.is_a?(ContextualLogger::LoggerMixin)
+        logger.error(error_message, **context)
+      else
+        logger.error(error_message)
```

Comment:
Hmm, maybe this whole interface could be conditional on `ContexualLogger`? It's unfortunate to accept `**context` and then discard it.

Reply:
The context is still put directly into the log message itself, so it's not entirely tossed.  And on top of that, since the logger can be changed at any point, the interface cannot change for contextual logging. We could make `escalate-contextual_logger_extention` to have only required with contextual logger is present, but that would end up looking just like this.  I'm all for removing the `contextual_logger` portion of this, it's just a shame to miss out on that context.

Reply:
That's what I have lined up in the current version of the code, `contextual_logger` has been move to the Gemfile and the code is using `defined?(ContextualLogger)` before checking `logger.is_a?(ContextualLogger::LoggerMixin)`